eyecite/annotate.py

```diff
@@ -0,0 +1,174 @@
+    to text_after[1:4] ("oo ") + text_after[8:10] ("ba").
+    """
+
+    def __init__(self, text_before, text_after):
+        """To set up, we need to populate self.offsets and self.updaters:
+            >>> SpanUpdater(text_before, text_after).offsets
+            [0, 4]
+            >>> SpanUpdater(text_before, text_after).updaters
+            [partial(shift_offset, delta=0), partial(shift_offset, delta=4)]
+        This indicates that offsets 0 to 4 need to be shifted by 0,
+        and offsets 4 and up need to be shifted by 4.
+        """
+        # helpers for the two kinds of updates we need to apply to offsets:
+        def shift_offset(offset, delta):
+            return offset + delta
+
+        def replace_offset(offset, new_offset):
+            return new_offset
+
+        # diff the two strings and set self.offsets and self.updaters:
+        offset = 0
+        delta = 0
+        self.offsets = offsets = []
+        self.updaters = updaters = []
+        for operation, amount in self.get_diff_steps(text_before, text_after):
+            if operation == "equal":
+                # start a new range with a relative delta,
+                # and push the offset forward
+                offsets.append(offset)
+                updaters.append(partial(shift_offset, delta=delta))
+                offset += amount
+            elif operation == "insert":
+                # push the delta forward
+                delta += amount
+            else:  # operation == 'delete'
+                # Start a new range with an absolute delta.
+                # Push the offset forward and delta backward.
+                offsets.append(offset)
+                updaters.append(
+                    partial(replace_offset, new_offset=offset + delta)
+                )
+                offset += amount
+                delta -= amount
+
+    @staticmethod
+    def get_diff_steps(a: str, b: str):
+        """Yield steps to turn a into b. Example:
+            >>> list(SpanUpdater.get_diff_steps("12 34 56", "12 78 34"))
+            [('equal', 3), ('insert', 3), ('equal', 2), ('delete', 3)]
+        Meaning: to turn a into b, keep the first 3 characters the same,
+        insert three new characters (we don't care what), keep the next
+        two characters, delete three characters.
+        """
+        diffs = SequenceMatcher(a=a, b=b, autojunk=False)
+        for operation, a1, a2, b1, b2 in diffs.get_opcodes():
+            if operation == "insert":
+                yield operation, b2 - b1
+            elif operation == "replace":
+                yield "delete", a2 - a1
+                yield "insert", b2 - b1
+            else:  # 'delete', 'equal'
+                yield operation, a2 - a1
+
+    # same as above using the diff-match-patch library, which may
+    # work better with some inputs because it offers minimal edit sequences:
+    # @staticmethod
+    # def get_dmp_diff_steps(self, a: str, b: str):
```

Comment:
Why is this commented out? Should you expose the option of using this method instead to the user?

Reply:
I can do that. My hunch is there's one right answer, and we'll find out which one shortly after we run this at scale. But I'm fine with having an option while we sort it out.

Reply:
Added a (for-now undocumented) option `annotate(use_dmp=True)`, defaulting to False. Using this option requires `pip install diff-match-patch`.